cmd/packageloader/loader/loader.go

```diff
@@ -0,0 +1,232 @@
+	imports := []string{}
+	fileList := []os.DirEntry{}
+	files, err := os.ReadDir(CURRENT_PATH)
+	if err != nil {
+		return err
+	}
+	for _, dirEntry := range files {
+		if dirEntry.Name()[len(dirEntry.Name())-2:] != "cx" {
+			continue
+		}
+		fileList = append(fileList, dirEntry)
+	}
+	if len(fileList) == 1 {
+		packageName, err := getPackageName(fileList[0])
+		if err != nil {
+			return err
+		}
+		imports, err = getImports(fileList[0], imports)
+		if err != nil {
+			return err
+		}
+		newPackage.PackageName = packageName
+	}
+	if len(fileList) > 1 {
+		samePackage := false
+		packageName := ""
+		samePackage, packageName, imports, err = comparePackageNames(fileList, imports)
+		if err != nil {
+			return err
+		}
+		if !samePackage {
+			log.Print("Files in directory " + CURRENT_PATH + " are not all in the same newPackage.\nSource of the error: " + packageName)
+			return errors.New("ErrMismatchedPackageFiles")
+		}
+		newPackage.PackageName = packageName
+	}
+	newPackage.addFiles(fileList)
+	packageList.hashPackage(&newPackage)
+
+	IMPORTED_DIRECTORIES = append(IMPORTED_DIRECTORIES, CURRENT_PATH)
+	for _, path := range imports {
+		if contains(SKIP_PACKAGES, path) {
+			continue
+		}
+		packageList.addPackagesIn(SRC_PATH + path)
+	}
+	return nil
+}
+
+// For a list of cx files, get their package names and return if they match, and add the imports
+func comparePackageNames(fileList []fs.DirEntry, imports []string) (bool, string, []string, error) {
+	packageName, err := getPackageName(fileList[0])
+	if err != nil {
+		return false, "", imports, err
+	}
+	imports, err = getImports(fileList[0], imports)
+	if err != nil {
+		return false, "", imports, err
+	}
+	for i := 1; i < len(fileList); i++ {
+		newPackageName, err := getPackageName(fileList[i])
+		if err != nil {
+			return false, "", imports, err
+		}
+		imports, err = getImports(fileList[i], imports)
+		if err != nil {
+			return false, "", imports, err
+		}
+		if newPackageName != packageName {
+			return false, newPackageName + "in" + fileList[i].Name(), imports, nil
+		}
+	}
+	return true, packageName, imports, nil
+}
+
+// Get the package name of a cx file
+func getPackageName(dirEntry fs.DirEntry) (string, error) {
+	file, err := os.Open(CURRENT_PATH + dirEntry.Name())
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+	scanner := bufio.NewScanner(file)
+	scanner.Split(bufio.ScanWords)
+
+	wordBefore := ""
+	for scanner.Scan() {
+		if scanner.Text() != "package" {
+			wordBefore = scanner.Text()
+			continue
+		}
+		if wordBefore == "//" {
+			wordBefore = scanner.Text()
+			continue
+		}
+		if scanner.Text() == "import" || scanner.Text() == "var" || scanner.Text() == "const" || scanner.Text() == "type" || scanner.Text() == "func" {
+			return "", errors.New("no package name found")
+		}
+		break
+	}
+	scanner.Scan()
+	return scanner.Text(), nil
+}
+
+// Get the import names in a cx file
+func getImports(dirEntry fs.DirEntry, imports []string) ([]string, error) {
```

Comment:
This is a `getImports()` so why do we need to give `imports []string` to it? Can we remove it?

Reply:
getImports() add the imports progressively as the program goes through the files, to avoid going through them twice and improve the speed of the program overall. If this function got all the imports at once, it would add an additional for loop.

Reply:
cant you just do `new imports, err:=getImport(dir)` then `importList+=newImports` ?